include/xtensor/xrepeat.hpp

```diff
@@ -0,0 +1,565 @@
+        void to_end(layout_type l);
+
+        template <class T>
+        simd_return_type<T> step_simd();
+
+        void step_leading();
+
+        template <class R>
+        void store_simd(const R& vec);
+
+    private:
+        S m_substepper;
+        const shape_type& m_shape;
+
+        std::ptrdiff_t m_repeating_steps;
+        std::vector<size_type> m_positions;
+        size_type m_subposition;
+
+        const size_type& m_repeating_axis;
+        const std::vector<std::ptrdiff_t>& m_repeats;
+
+        void make_step(size_type dim, size_type n);
+        void make_step_back(size_type dim, size_type n);
+
+        std::vector<size_type> get_next_positions(const size_type dim, const size_type steps_to_go) const;
+        std::vector<size_type> get_next_positions_back(const size_type dim, const size_type steps_to_go) const;
+    };
+
+    template <class CT>
+    class xrepeat :
+        public xiterable<xrepeat<CT>>,
+        public xaccessible<xrepeat<CT>>
+    {
+    public:
+        using xexpression_type = std::decay_t<CT>;
+        using value_type = typename xexpression_type::value_type;
+        using shape_type = typename xexpression_type::shape_type;
+
+        using container_type = xcontainer_inner_types<xrepeat<CT>>;
+        using reference = typename container_type::reference;
+        using const_reference = typename container_type::const_reference;
+        using size_type = typename container_type::size_type;
+        using temporary_type = typename container_type::temporary_type;
+
+        static constexpr layout_type static_layout = xexpression_type::static_layout;
+        using bool_load_type = typename xexpression_type::bool_load_type;
+        using pointer = typename xexpression_type::pointer;
+        using const_pointer = typename xexpression_type::const_pointer;
+        using difference_type = typename xexpression_type::difference_type;
+
+        using iterable_type = xiterable<xrepeat<CT>>;
+        using stepper = typename iterable_type::stepper;
+        using const_stepper = typename iterable_type::stepper;
+
+        template<class CTA>
+        explicit xrepeat(CTA&& e, std::vector<std::ptrdiff_t>&& repeats, size_type axis);
+
+        template <class... Args>
+        reference operator()(Args... args);
+
+        template <class... Args>
+        const_reference operator()(Args... args) const;
+
+        const shape_type& shape() const noexcept;
+
+        template <class It>
+        reference element(It first, It last);
+
+        template <class It>
+        const_reference element(It first, It last) const;
+
+        stepper stepper_begin() const;
+
+        stepper stepper_begin(const shape_type& s) const;
+
+        stepper stepper_end(const layout_type l) const;
+
+        stepper stepper_end(const shape_type& s, const layout_type l) const;
+
+    private:
+        CT m_e;
+        const size_type m_repeating_axis;
+        const std::vector<std::ptrdiff_t> m_repeats;
```

Comment:
If you go with the overladed API and the implementation function accepting a universal reference, this type should be changed for `xtl::const_closure_type_t<E>`

Reply:
Thanks for the useful feedback. I hope I understood you correctly. Otherwise feel free to give me more feedback.